battenberg_gitlab/cli.py

```diff
@@ -5,39 +5,122 @@
 
 
-logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger('battenberg_gitlab')
+# Ensure we always receive debug messages.
+handler = logging.StreamHandler()
+handler.setLevel(logging.INFO)
+logger.addHandler(handler)
 
 
-def main(config_file: str = None,
-         gitlab_server: str = None,
-         group_ids: List[int] = None,
-         # Expect filters to be in "<filter type>=<keyword>" format.
-         cli_filters: List[str] = None,
-         workspace_path: str = None,
-         checkout: str = None):
+@click.group()
+@click.option(
+    '--config-file',
+    default=None,
+    help='Where to read the python-gitlab configuration from.',
+    type=click.Path(exists=True, file_okay=True, dir_okay=False)
+)
+@click.option(
+    '--gitlab-id',
+    default=None,
+    help='Where to read the python-gitlab configuration from.',
+    type=click.STRING
+)
+@click.option(
+    '--verbose',
+    default=False,
+    is_flag=True,
+    help='Enables the debug logging.'
+)
+@click.pass_context
+def main(ctx, config_file: str, gitlab_id: str, verbose: bool):
+    """
+    \f
 
-    # TODO ZOMLINFRA-480 Add in click CLI.
-    # TODO ZOMLINFRA-480 Stop hard coding the ZO & ZOML specific stuff.
-    gitlab_server = 'zgtools'
-    group_ids = [109, 108, 106]  # ZOML Gitlab groups
-    cli_filters = ['tag=archetype.py-ml']
-    checkout = 'v1.1.0'
+    Script entry point for Battenberg-gitlab commands.
+    """
+    ctx.obj = dict()
+    ctx.obj.update({
+        'config_file': config_file,
+        'gitlab_id': gitlab_id,
+        'verbose': verbose
+    })
 
-    project_filters = parse_project_filters(cli_filters)
+    level = logging.DEBUG if verbose else logging.INFO
+    logger.setLevel(level)
 
-    gl = init_gitlab(gitlab_server, config_file)
+
+@main.command()
+@click.option(
+    '--workspace',
+    help='Where to locally clone found repos to in the local filesystem. Defaults to /tmp',
+    default=None,
+    type=click.Path(exists=False)
+)
+@click.option(
+    '--group-name',
+    help='Name of the Gitlab group to .',
+    type=click.STRING,
+    multiple=True
+)
+@click.option(
+    '--project-filter',
+    help='Expected to be in "<filter type>=<keyword>" format. Supported types: "tag".',
+    type=click.STRING,
+    multiple=True
+)
+@click.pass_context
+def search(ctx, workspace: str, group_name: List[str], project_filter: List[str]):
+    gl = init_gitlab(ctx.obj['gitlab_id'], ctx.obj['config_file'])
+    project_filters = parse_project_filters(project_filter)
 
     try:
-        apply(gl, group_ids, project_filters, workspace_path, checkout)
+        projects = search_projects(gl, group_name, project_filters, workspace)
     except ProjectNotFoundError:
-        print(f'No projects found matching filters: {cli_filters}')
+        logger.info(f'No projects found matching filters: {project_filter}')
+        return
+
+    for project, version in projects.items():
+        logger.info(f'{project.name}: {version}')
```

Comment:
Currently as we haven't actioned [ZOMLINFRA-426](https://zbrt.atl.zillow.net/browse/ZOMLINFRA-426) the `version` for every repo will just be `Prepared template upgrade`, see [here](https://gitlab.zgtools.net/zillow-offers/zoml/ml-infrastructure/example-project/commits/template) for example. If we make the commit messages for install & upgrade customizable then we can set them in `battenberg-gitlab apply` in such a way that `battenberg-gitlab search` can interpret them 😄